source/common/runtime/runtime_impl.cc

```diff
@@ -23,6 +23,29 @@ namespace Runtime {
 
 const size_t RandomGeneratorImpl::UUID_LENGTH = 36;
 
+uint64_t RandomGeneratorImpl::random() {
+  static thread_local uint8_t buffered[2048];
+  static thread_local size_t buffered_idx = sizeof(buffered);
+
+  // Prefetch 2048 bytes of randomness. buffered_idx is initialized to sizeof(buffered),
+  // i.e. out-of-range value, so the buffer will be filled with randomness on the first
+  // call to this function.
+  if (buffered_idx >= sizeof(buffered)) {
+    int rc = RAND_bytes(buffered, sizeof(buffered));
+    ASSERT(rc == 1);
+    UNREFERENCED_PARAMETER(rc);
+    buffered_idx = 0;
+  }
+
+  // Consume sizeof(uint64) bytes from the buffer.
+  ASSERT(buffered_idx + sizeof(uint64_t) <= sizeof(buffered));
+  uint64_t rand;
+  std::memcpy(reinterpret_cast<void*>(&rand), &buffered[buffered_idx], sizeof(uint64_t));
```

Comment:
Nit: is this really more readable or faster than a simple assignment?

Reply:
This is the correct way to do it, iirc, doing otherwise is flirting with undefined behavior. The compiler will optimize out the memcpy.

Reply:
If you make `buffered` an array of `uint64_t`, there is absolutely no undefined behavior.

Reply:
I agree with @htuch .. Why can't we totally avoid the copy, and just have a single integer to track the array consumption, and `return buffered[idx]` ? Is there some compiler issue at play here? And if so, (another uninitiated question), is it bad to have a global (dynamically allocated array) of these integer indices/and arrays, [one per thread], and use the thread specific slot? Cache misses wont be bad as long as the thread stays on the same core.

Reply:
@rshriram a global array of indices could be less efficient since the indices for different threads would be on the same cache line (at least on modern x86) so the cache line would bounce between cores unnecessarily.

Reply:
Yes, please just keep is simple. Use a thread local array of `uint64_t` which will be properly aligned, then just return the current entry and advance without a local copy.

Reply:
Ah shoot yes.. the index array would end up causing lots of false sharing. But the question of 2k vs 4k vs 8k byte array size remains..

Reply:
Duh! Array of `uint64_t` is indeed a better choice. Fixed.